src/stm32f1/pwm.cpp

```diff
@@ -0,0 +1,230 @@
+#include "pin.hpp"
+#include "power.hpp"
+#include "pwm_reg.hpp"
+#include <cmath>
+#include <libhal-arm-mcu/stm32f1/clock.hpp>
+#include <libhal-arm-mcu/stm32f1/constants.hpp>
+#include <libhal-arm-mcu/stm32f1/pin.hpp>
+#include <libhal-arm-mcu/stm32f1/pwm.hpp>
+#include <libhal-util/bit.hpp>
+
+namespace hal::stm32f1 {
+namespace {
+[[nodiscard]] pwm_reg_t* get_pwm_reg(peripheral p_id)
+{
+  if (p_id == peripheral::timer1) {
+    return pwm_timer1;
+  }
+  return pwm_timer8;
+}
+[[nodiscard]] peripheral get_peripheral_id(pwm_pins p_pin)
+{
+  if (p_pin == pwm_pins::pa8 || p_pin == pwm_pins::pa9 ||
+      p_pin == pwm_pins::pa10 || p_pin == pwm_pins::pa11) {
+
+    return peripheral::timer1;
+  } else {
+    hal::safe_throw(hal::operation_not_supported(nullptr));
+  }
+  return peripheral::timer8;
+}
+
+[[nodiscard]] float get_duty_cycle(pwm_pins p_pin,
+                                   uint32_t volatile* compare_register)
+{
+  // ccr / arr
+  auto peripheral_id = get_peripheral_id(p_pin);
```

Comment:
```suggestion
  auto const peripheral_id = get_peripheral_id(p_pin);
```

Try and make variables const whenever possible.